### PR TITLE
More contention reductions

### PR DESCRIPTION
**Pull request by pettyjamesm**
Continuation of contention reducing changes in https://github.com/prestosql/presto/pull/6097 extracted from https://github.com/prestodb/presto/pull/15509

Seven commits:
- Cache `ListenableFuture<?> isBlocked` in various operators to avoid repeatedly contending for locks and potentially registering new callbacks once those operators are blocked
- Add unsynchronized fast-paths in ExchangeClient to avoid unnecessary contention
- Refactor `ExchangeClient#pollPage` to avoid redundant checks and synchronizations
- Use `AtomicIntegerFieldUpdater` for local exchange `PageReference`
- Reduce synchronization in `LocalExchangeMemoryManager` by making unblocking an edge-triggered event that only a single caller initiates.
- Add unsynchronized common-case fast paths in `LocalExchangeSource`
- Remove synchronized for final field read in `MemoryPool#getMaxBytes()`

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java [hunk 1]
```diff
@@ -261,8 +262,13 @@ public boolean isFinished()
     @Override
     public ListenableFuture<?> isBlocked()
     {
-        ListenableFuture<?> blocked = partitionFunction.isFull();
-        return blocked.isDone() ? NOT_BLOCKED : blocked;
+        if (isBlocked.isDone()) {
+            isBlocked = partitionFunction.isFull();
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java:266]: future for `partitionFunction` is already cached

File: presto-main/src/main/java/io/prestosql/operator/TaskOutputOperator.java [hunk 2]
```diff
@@ -122,8 +123,13 @@ public boolean isFinished()
     @Override
     public ListenableFuture<?> isBlocked()
     {
-        ListenableFuture<?> blocked = outputBuffer.isFull();
-        return blocked.isDone() ? NOT_BLOCKED : blocked;
+        if (isBlocked.isDone()) {
+            isBlocked = outputBuffer.isFull();
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/TaskOutputOperator.java:127]: future of output buffer is already cached

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSinkOperator.java [hunk 3]
```diff
@@ -128,7 +129,13 @@ public boolean isFinished()
     @Override
     public ListenableFuture<?> isBlocked()
     {
-        return sink.waitForWriting();
+        if (isBlocked.isDone()) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSinkOperator.java:132]: future of sink is already cached

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSourceOperator.java [hunk 4]
```diff
@@ -103,7 +104,13 @@ public boolean isFinished()
     @Override
     public ListenableFuture<?> isBlocked()
     {
-        return source.waitForReading();
+        if (isBlocked.isDone()) {
+            isBlocked = source.waitForReading();
+            if (isBlocked.isDone()) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSourceOperator.java:109]: future of source is already cached

File: presto-main/src/main/java/io/prestosql/operator/ExchangeOperator.java [hunk 5]
```diff
@@ -92,6 +92,7 @@ public void noMoreOperators()
     private final PlanNodeId sourceId;
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/ExchangeOperator.java:92]: Change commit name `Cache ExchangeClient#isBlocked` future. Add explanation why `io.prestosql.operator.ExchangeClient#isBlocked` creates a new future every time (because they can be cancelled in `io.prestosql.server.protocol.Query#waitForResults`)

File: presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java (hunk 1 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java:266]: If you mean the cached future in `OutputBufferMemoryManager` then you're technically correct, except that retrieving that instance is through `OutputBufferMemoryManager#getBufferBlockedFuture()` which needs to synchronize on the heavily lock on each access.

This is especially painful since each thread working through the same task's driver loop necessarily shares the output buffer instance and will each call 3 different methods in rapid succession that all check whether the operator is currently blocked (`isBlocked()`, `needsInput()`, `isFinished()`). This is on top of:
- HTTP threads decrementing memory when acknowledging pages
- other worker threads incrementing memory after enqueuing pages
- Memory tracker future callbacks
- Constant checks against `OutputBufferMemoryManager#isOverutilized()` task concurrency levels.
- The fact that [this call into the system memory context](https://github.com/prestosql/presto/blob/37c74ad15f5f0c9c770c2aa491e49ffe5146d4b1/presto-main/src/main/java/io/prestosql/execution/buffer/OutputBufferMemoryManager.java#L91) is in effect entering a single global lock tree for every memory tracker update in any operator on any thread anywhere, which means that a thread holding the output buffer monitor can and does often block trying to acquire the memory tracking lock which subsequently locks out all threads calling any synchronized method on the output buffer.

This change _somewhat_ reduces the pain of all of that by caching the future in the operator itself when blocked so that subsequent checks don't re-enter the same massive bottleneck, but offers no improvement when not blocked. I'm in the process of putting together a proprosal to address the memory tracking and output buffer issues now.

File: presto-main/src/main/java/io/prestosql/operator/TaskOutputOperator.java (hunk 2 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/TaskOutputOperator.java:127]: Same justification as [my reply](https://github.com/prestosql/presto/pull/6260#discussion_r540921443) about `PartitionedOutputOperator`

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSinkOperator.java (hunk 3 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSinkOperator.java:132]: Slightly different context, but essentially is the same justification as [my reply](https://github.com/prestosql/presto/pull/6260#discussion_r540921443) for `PartitionedOutputOperator`, except with `LocalExchangeMemoryManager` instead of `OutputBufferMemoryManager`. Admittedly, the problem is not nearly as bad there because it doesn't call into the memory tracking system but synchronizing on each access is still unnecessary.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSourceOperator.java (hunk 4 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSourceOperator.java:109]: Similar justification here as for `LocalExchangeSinkOperator`.

File: presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java:266]: Please add a comment in code before `if (isBlocked.isDone()) {`

Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/PartitionedOutputOperator.java:267]: Added

File: presto-main/src/main/java/io/prestosql/operator/exchange/BroadcastExchanger.java [hunk 6]
```diff
@@ -27,19 +28,23 @@
 {
     private final List<Consumer<PageReference>> buffers;
     private final LocalExchangeMemoryManager memoryManager;
+    private final LongConsumer onPageReleased;
+    private final int bufferCount;
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/BroadcastExchanger.java:32]: I don't think this buys anything. I would continue to use `buffers.size()`

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java [hunk 7]
```diff
@@ -17,24 +17,28 @@
...
+    private static final AtomicIntegerFieldUpdater<PageReference> REFERENCE_COUNT_UPDATER = AtomicIntegerFieldUpdater.newUpdater(PageReference.class, "referenceCount");
+
     private final Page page;
-    private final Runnable onFree;
-    private final AtomicInteger referenceCount;
+    private final LongConsumer onPageReleased;
+    private volatile int referenceCount;
 
-    public PageReference(Page page, int referenceCount, Runnable onFree)
+    public PageReference(Page page, int referenceCount, LongConsumer onPageReleased)
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:36]: Changing from `Runnable` to `LongConsumer` doesn't improve anything. Undo. Alternatively make it `Consumer<Page>` as it's more explicit than generic long.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java [hunk 8]
```diff
@@ -17,24 +17,28 @@
...
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.LongConsumer;
 
+import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 @ThreadSafe
-class PageReference
+final class PageReference
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:28]: unrelated change. Extract to separate commit

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java [hunk 9]
```diff
@@ -44,11 +48,26 @@ public long getRetainedSizeInBytes()
...
+        int referenceCount = REFERENCE_COUNT_UPDATER.decrementAndGet(this);
         checkArgument(referenceCount >= 0, "Page reference count is negative");
         if (referenceCount == 0) {
-            onFree.run();
+            onPageReleased.accept(page.getRetainedSizeInBytes());
         }
         return page;
     }
+
+    @Override
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:59]: extract to separate commit

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java [hunk 10]
```diff
@@ -44,11 +48,26 @@ public long getRetainedSizeInBytes()
...
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("size", getRetainedSizeInBytes())
+                .add("referenceCount", referenceCount)
+                .toString();
+    }
+
+    public static LongConsumer releaseListenerForExchangeMemoryManager(LocalExchangeMemoryManager memoryManager)
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:68]: this is binding `PageReference` code with `LocalExchangeMemoryManager` unnecessarily.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java [hunk 11]
```diff
@@ -17,24 +17,28 @@
...
 
+import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
 @ThreadSafe
-class PageReference
+final class PageReference
 {
+    private static final AtomicIntegerFieldUpdater<PageReference> REFERENCE_COUNT_UPDATER = AtomicIntegerFieldUpdater.newUpdater(PageReference.class, "referenceCount");
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:30]: I don't think `AtomicInteger` is a problem. Page granularity is not such a big deal. If there was `AtomicInteger` per row that would be a problem

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java [hunk 12]
```diff
@@ -47,31 +49,37 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
...
-
-            // if we are full, then breakout
-            if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone()) {
-                return;
+        long bufferedBytes = this.bufferedBytes.addAndGet(bytesAdded);
+        // detect the transition from above to below the full boundary
+        if (bufferedBytes <= maxBufferedBytes && (bufferedBytes - bytesAdded) > maxBufferedBytes) {
+            SettableFuture<?> future;
+            synchronized (this) {
+                //  if we have no callback waiting, return early
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:57]: nit: extra space after `//`

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java [hunk 13]
```diff
@@ -47,31 +49,37 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
...
+                if (notFullFuture == NOT_FULL) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:58]: use `notFullFuture.isDone()`

Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:58]: recheck `if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone())`

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java [hunk 14]
```diff
@@ -47,31 +49,37 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
...
-        // if we are full and the current not full future is already complete, create a new one
-        if (bufferedBytes.get() > maxBufferedBytes && notFullFuture.isDone()) {
-            notFullFuture = SettableFuture.create();
+        if (bufferedBytes.get() <= maxBufferedBytes) {
+            return NOT_BLOCKED;
+        }
+        SettableFuture<?> future;
+        synchronized (this) {
+            // if we are full and the current not full future is already complete, create a new one
+            if ((notFullFuture == NOT_FULL || notFullFuture.isDone()) && bufferedBytes.get() > maxBufferedBytes) {
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:77]: simplify:
```
// optionally
if (bufferedBytes.get() <= maxBufferedBytes) {
  return NOT_BLOCKED;
}
if (notFullFuture.isDone()) {
  notFullFuture = SettableFuture.create();
}
```

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java [hunk 15]
```diff
@@ -22,10 +22,12 @@
 import java.util.concurrent.atomic.AtomicLong;
 
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
 
 @ThreadSafe
 public class LocalExchangeMemoryManager
 {
+    private static final ListenableFuture<?> NOT_BLOCKED = immediateFuture(null);
     private static final SettableFuture<?> NOT_FULL;
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:31]: let's remove this. This makes he code unnecessarily complicated.

**Comment by sopel39** [comment]
reviewed up to `Reduce synchronization in LocalExchangeMemoryManager`

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 7 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:36]: `Runnable` requires that each `PageReference` instance is initialized with a lambda that captures state from the creating context in order to perform cleanup. a single LongConsumer can be created per exchange sink and used for all created instances.

`Consumer<Page>` might be ok, but also opens the door to accidental ambiguity like whether `Page#getRetainedSizeInBytes()` or `Page#getSizeInBytes()` should be used as the decrement.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 8 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:28]: It's material to the change to using `AtomicIntegerFieldUpdater` which will not work as expected with subclasses.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 11 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:30]: It's still an extra allocation (with object headers and alignment overhead) and extra pointer indirection at usage time. The improvement is minor but enough of these are created through local exchanges that embedding the counter is worthwhile.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 9 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:59]: This was a bit of a hack, without adding this toString() implementation the `referenceCount` field is detected as unused and fails the build.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 10 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:68]: This avoids accidentally creating a megamorphic call site inside of `removePage()`. Currently, there are only two implementations of used- this one and the anonymous one created by `PassthroughExchanger`. If each site created their own lambda it would be megamorphic.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 13 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:58]: It's necessary to *not* re-check inside of the synchronized block in order to preserve existing semantics. Re-checks can race with concurrent modifications whereas previously all modifications were serialized through the object monitor. The effect was that previously the transition edge from above-full to below full always unblocked any waiters, and not doing so here (even if we've transitioned back up above the full boundary in the mean time) is a departure from that behavior.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 10 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:68]: This avoids accidentally creating a megamorphic call site inside of `removePage()`. Currently, there are only two implementations used- this one and the anonymous one created by `PassthroughExchanger`. If each exchanger created their own lambda it would be megamorphic.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 13 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:58]: `NOT_FULL` is the sentinel value for "no listeners". Any other value is a future that might not be done, but that check can occur outside of the synchronized block and without stalling to do a volatile read against that future's `isDone()` state. Redudant `set(null)` calls are benign but this implementation preserves the ability to alias check the `NOT_FULL` sentinel to detect listener registration.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 14 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java]: Simplified, also realized that the `notFullFuture.isDone()` check is not correct as it can race with the callback. Updated to now only use the  `notFullFuture == NOT_DONE` alias check for callback registration.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 15 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java]: After migrating all checks against the `NOT_FULL` sentinel, it became clear that `null` would be a more clear fit. I've changed it to avoid having `NOT_FULL` and using null to represent no callback registered instead.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 13 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java]: Replaced with null check instead.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 14 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java]: Updated again to use a null sentinel and this becomes a null check.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 10 above)
Comment (erichwang) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:68]: This is certainly an interesting idea, but does this actually make a difference? Presto is pretty much built around pushing megamorphic call sites to page-sized granularities. If this needs to be specialized like this, we probably have much bigger issues throughout the entire engine.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java [hunk 16]
```diff
@@ -19,41 +19,37 @@
...
-    @GuardedBy("lock")
-    private SettableFuture<?> notEmptyFuture = NOT_EMPTY;
+    @Nullable
+    @GuardedBy("this")
+    private SettableFuture<?> notEmptyFuture; // null indicates no callback has been registered
 
-    @GuardedBy("lock")
-    private boolean finishing;
+    @GuardedBy("this")
+    private final AtomicBoolean finishing = new AtomicBoolean(false);
```
Comment (erichwang) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java:52]: the default constructor value for AtomicBoolean is already false.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java [hunk 17]
```diff
@@ -19,41 +19,37 @@
...
-    static {
-        NOT_EMPTY = SettableFuture.create();
-        NOT_EMPTY.set(null);
-    }
+    private static final ListenableFuture<?> NOT_BLOCKED = immediateFuture(null);
 
     private final Consumer<LocalExchangeSource> onFinish;
 
-    private final BlockingQueue<PageReference> buffer = new LinkedBlockingDeque<>();
+    private final LinkedBlockingDeque<PageReference> buffer = new LinkedBlockingDeque<>();
```
Comment (erichwang) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java:44]: Why do we need to change BlockingQueue to LinkedBlockingDeque?

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java [hunk 18]
```diff
@@ -164,18 +178,20 @@ public void finish()
...
             notEmptyFuture = this.notEmptyFuture;
-            this.notEmptyFuture = NOT_EMPTY;
+            this.notEmptyFuture = null;
         }
 
         // notify readers outside of lock since this may result in a callback
-        notEmptyFuture.set(null);
+        if (notEmptyFuture != null) {
+            notEmptyFuture.set(null);
+        }
```
Comment (erichwang) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java:194]: Any reason making notEmptyFuture nullable is better than what we had before? This seems bulkier without notable benefit?

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 7 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:36]: > Runnable requires that each PageReference instance is initialized with a lambda that captures state from the creating context in order to perform cleanup. a single LongConsumer can be created per exchange sink and used for all created instances.

You could make it `Consumer<PageReference>`
Then lambda can be:
```
reference -> doSomething(reference.getRetainedSizeInBytes()
```
It's cleaner and more explicit than generic long.

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 9 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:59]: You could add `toString` without `referenceCount` in previous commit and then add `referenceCount` to `toString` in this commit

File: presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java (hunk 10 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java:68]: I agree with @erichwang . We should not obscure readability for negligible perf improvements

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java [hunk 19]
```diff
@@ -16,28 +16,26 @@
...
-    }
+    private static final ListenableFuture<?> NOT_BLOCKED = immediateFuture(null);
 
     private final long maxBufferedBytes;
     private final AtomicLong bufferedBytes = new AtomicLong();
 
+    @Nullable
     @GuardedBy("this")
-    private SettableFuture<?> notFullFuture = NOT_FULL;
+    private SettableFuture<?> notFullFuture; // null represents "no callback registered"
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:38]: please initialize it with `NOT_BLOCKED`

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java [hunk 20]
```diff
@@ -47,31 +45,39 @@ public LocalExchangeMemoryManager(long maxBufferedBytes)
...
-
-            // if we are full, then breakout
-            if (bufferedBytes.get() > maxBufferedBytes || notFullFuture.isDone()) {
-                return;
+        long bufferedBytes = this.bufferedBytes.addAndGet(bytesAdded);
+        // detect the transition from above to below the full boundary
+        if (bufferedBytes <= maxBufferedBytes && (bufferedBytes - bytesAdded) > maxBufferedBytes) {
+            SettableFuture<?> future;
+            synchronized (this) {
+                // if we have no callback waiting, return early
```
Comment (sopel39) [inline presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:53]: This optimization reduces readability of concurrent code for little benefit. In practice, we already synchronized on `this`, so doing extra checks like `notFullFuture.isDone()` is not a big issue. Also, we should recheck that buffer is still not full before triggering future. That check is cheap, but waking up drivers to do empty processing is not.

> The effect was that previously the transition edge from above-full to below full always unblocked any waiters, and not doing so here (even if we've transitioned back up above the full boundary in the mean time) is a departure from that behavior.

We cold not transition back to above in the meantime because all methods in `LocalExchangeMemoryManager` are synchronized. We could transition to above immediately after future was set, but that's slightly different behavior.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java (hunk 16 above)
Comment (sopel39) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java:52]: I would just use `volatile boolean finishing`

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java (hunk 17 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java:44]: Accidentally left this behind, I changed it because it was easier to navigate into the implementation of `LinkedBlockingDeque` methods in my IDE if the field was declared with the exact type. Will revert it to the previous declaration.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java (hunk 16 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java:52]: Sure, will make this a `volatile boolean` instead.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java (hunk 18 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeSource.java:194]: Compared to what was happening before, this is indeed better in terms of performance: 
- Overwriting `this.notEmptyFuture` unconditionally is an additional (potentially irrelevant) store operation that must be flushed to memory as part of releasing the monitor. Skipping the store if the field is already null (ie: no listener is registered) avoids that.
- Unconditionally calling `SettableFuture#set(null)` in the previous implementation required a volatile CAS even when no listener was registered which performs poorly compared to a null check (which are exceedingly cheap)

I think the previous use of having the `static final SettableFuture<?> NOT_EMPTY` constant was arguable _less_ clear about what was going on. Conceptually, this field has two states: listener registered and no listener registered. You could make them _appear_ the same by using the constant, but then other parts of code are always checking `isDone()` which is expensive (volatile read) and in my opinion creates a hazard for race conditions since `set(null)` can race with the `isDone()` checks without paying close attention to subtle details of what happens inside of the syncronized sections.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 19 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:38]: `NOT_BLOCKED` is not a `SettableFuture<?>` but rather a `ListenableFuture<?>` which can't be set (it's already completed). See https://github.com/prestosql/presto/pull/6260#discussion_r547368980 for some additional detail around this implementation strategy.

File: presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java (hunk 20 above)
Comment (pettyjamesm) [reply presto-main/src/main/java/io/prestosql/operator/exchange/LocalExchangeMemoryManager.java:53]: > This optimization reduces readability of concurrent code...

I disagree. The prior implementation seemed "simpler" in that there was no null checking, but it was also masking a subtle problem: checking `isDone()` while holding the lock can race with `future.set(null)` which occurs _outside_ of the synchronized section. This happens to be _ok_ because the field is overwritten inside the synchronized section, but is much more subtle than a null check IMO.

> ... so doing extra checks like `notFullFuture.isDone()` is not a big issue

It's significantly more expensive than a null check because it requires a volatile read (while holding the lock) of the value at the other end of the current reference, meaning that having acquired the object monitor is insufficient to have brought that memory into the CPU cache and will _at best_ need to read it from a shared CPU cache if not from main memory.

Now, will the extra cycles incurred in doing that dominate performance? Not likely, but there is an Amdahl's law component to the duration that any thread spends in the synchronized section of `updateMemoryUsage(long)` or `getNotFullFuture()`.

> That check is cheap, but waking up drivers to do empty processing is not.

I agree, the question is whether changing the semantics are a good idea as part of this refactor. Previously, all modifications were linearized by the object monitor. Now, they linearize on the `bufferedBytes` atomic to accomplish the same goal and the synchronized block is establishing a consistent view of the callback field. I can't say how many workloads today depend on "fluttering" against the full vs not-full state and whether they would all benefit from having some of those unblocking notifications suppressed. In the interest of being conservative, I would prefer to keep the previous semantics (but with less lock contention) than potentially regress performance.

**Review by dain — APPROVED** [review]
Looking good

File: core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchangeSource.java [hunk 21]
```diff
@@ -30,30 +31,24 @@
...
 @ThreadSafe
 public class LocalExchangeSource
 {
-    private static final SettableFuture<?> NOT_EMPTY;
-
-    static {
-        NOT_EMPTY = SettableFuture.create();
-        NOT_EMPTY.set(null);
-    }
+    private static final ListenableFuture<?> NOT_BLOCKED = immediateFuture(null);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchangeSource.java:40]: Under the hood it’s already a static singleton (which is not publicly exported) so between adding a new static final reference to this class or statically importing that same constant  reference from, say, the Operator interface it seems more straightforward duplicate the constant reference here.

File: core/trino-main/src/main/java/io/trino/operator/ExchangeClient.java [hunk 22]
```diff
@@ -322,52 +323,73 @@ public synchronized void scheduleRequestIfNecessary()
...
+            systemMemoryContext.setBytes(bufferRetainedSizeInBytes);
+            successfulRequests++;
+            // AVG_n = AVG_(n-1) * (n-1)/n + VALUE_n / n
+            averageBytesPerRequest = (long) (1.0 * averageBytesPerRequest * (successfulRequests - 1) / successfulRequests + responseSize / successfulRequests);
 
-        bufferRetainedSizeInBytes += pagesRetainedSizeInBytes;
-        maxBufferRetainedSizeInBytes = Math.max(maxBufferRetainedSizeInBytes, bufferRetainedSizeInBytes);
-        systemMemoryContext.setBytes(bufferRetainedSizeInBytes);
-        successfulRequests++;
+            if (!pages.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/ExchangeClient.java:368]: I think you’re on to something here. Comments elsewhere suggest that updating the per-request and request count stars must occur even when no pages are added, but there is a real opportunity to skip the memory tracker update of 0 bytes (huge contention bottleneck, although there might be more early returns on 0) and the buffer lock while passing an empty list argument to addAll.

Might be

[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/operator/ExchangeClient.java]: Reorganized the method to only call `pageBuffer.addAll` and update system memory when the pages list is not empty.
